src/Confluent.Kafka/Impl/LibRdKafka.cs

```diff
@@ -967,56 +967,42 @@ internal static ErrorCode committed(IntPtr rk, IntPtr partitions, IntPtr timeout
         internal static ErrorCode position(IntPtr rk, IntPtr partitions)
             => _position(rk, partitions);
 
-        /// <summary>
-        ///     Var-arg tag types, used in producev
-        /// </summary>
-        internal enum ProduceVarTag
+        private delegate IntPtr Produceva(IntPtr rk,
+            rd_kafka_vu[] vus,
+            IntPtr size);
+        
+        private static Produceva _produceva;
+
+        internal static ErrorCode produceva(
+            IntPtr rk,
+            string topic,
+            int partition,
+            IntPtr msgflags,
+            IntPtr val, UIntPtr len,
+            IntPtr key, UIntPtr keylen,
+            long timestamp,
+            IntPtr headers,
+            IntPtr msg_opaque)
         {
-            End,       // va-arg sentinel
-            Topic,     // (const char *) Topic name
-            Rkt,       // (rd_kafka_topic_t *) Topic handle
-            Partition, // (int32_t) Partition
-            Value,     // (void *, size_t) Message value (payload)
-            Key,       // (void *, size_t) Message key
-            Opaque,    // (void *) Application opaque
-            MsgFlags,  // (int) RD_KAFKA_MSG_F_.. flags
-            Timestamp, // (int64_t) Milliseconds since epoch UTC
-            Header,    // (const char *, const void *, ssize_t) Message Header
-            Headers,   // (rd_kafka_headers_t *) Headers list
+            using (var topicStrPinned = new Util.Marshal.StringAsPinnedUTF8(topic))
+            {
+                rd_kafka_vu[] vus =
+                {
+                    new rd_kafka_vu() {vt = rd_kafka_vtype.Topic,     data  = new vu_data() {topic = topicStrPinned.Ptr}},
+                    new rd_kafka_vu() {vt = rd_kafka_vtype.Partition, data  = new vu_data() {partition = partition}},
+                    new rd_kafka_vu() {vt = rd_kafka_vtype.MsgFlags,  data  = new vu_data() {msgflags = msgflags}},
+                    new rd_kafka_vu() {vt = rd_kafka_vtype.Value,     data  = new vu_data() {val = new ptr_and_size() {data = val, size = len}}},
+                    new rd_kafka_vu() {vt = rd_kafka_vtype.Key,       data  = new vu_data() {val = new ptr_and_size() {data = key, size = keylen}}},
```

Comment:
this will work, but you should assign to `key` here, not `val`.

Reply:
My bad, done